torch/csrc/api/src/nn/modules/conv.cpp

```diff
@@ -20,119 +25,163 @@ ConvImpl<D, Derived>::ConvImpl(const ConvOptions<D>& options_)
+        .padding(0)
+        .dilation(options.dilation())
+        .groups(options.groups()));
   }
-  return torch::conv1d(
-      input,
-      weight,
-      bias,
-      options.stride(),
-      options.padding(),
-      options.dilation(),
-      options.groups());
+  return F::conv1d(
+    input,
+    weight,
+    bias,
+    Conv1dOptions()
+      .stride(options.stride())
+      .padding(options.padding())
+      .dilation(options.dilation())
+      .groups(options.groups()));
 }
 
+Conv2dImpl::Conv2dImpl(
+    ConvOptions<2> options_)
+    : ConvImpl(std::move(options_).transposed(false).output_padding(0)) {}
+
 Tensor Conv2dImpl::forward(const Tensor& input) {
-  if (options.transposed()) {
-    return torch::conv_transpose2d(
-        input,
-        weight,
-        bias,
-        options.stride(),
-        options.padding(),
-        options.output_padding(),
-        options.groups(),
-        options.dilation());
+  if (c10::get_if<enumtype::kCircular>(&options.padding_mode())) {
+    std::vector<int64_t> expanded_padding = {
+      ((*options.padding())[1] + 1) / 2, (*options.padding())[1] / 2,
+      ((*options.padding())[0] + 1) / 2, (*options.padding())[0] / 2};
+    return F::conv2d(
+      F::pad(input, PadOptions(expanded_padding).mode(torch::kCircular)),
+      weight, bias,
+      Conv2dOptions()
+        .stride(options.stride())
+        .padding(0)
+        .dilation(options.dilation())
+        .groups(options.groups()));
   }
-  return torch::conv2d(
-      input,
-      weight,
-      bias,
-      options.stride(),
-      options.padding(),
-      options.dilation(),
-      options.groups());
+  return F::conv2d(
+    input,
+    weight,
+    bias,
+    Conv2dOptions()
+      .stride(options.stride())
+      .padding(options.padding())
+      .dilation(options.dilation())
+      .groups(options.groups()));
 }
 
+Conv3dImpl::Conv3dImpl(
+    ConvOptions<3> options_)
+    : ConvImpl(std::move(options_).transposed(false).output_padding(0)) {}
+
 Tensor Conv3dImpl::forward(const Tensor& input) {
-  if (options.transposed()) {
-    return torch::conv_transpose3d(
-        input,
-        weight,
-        bias,
-        options.stride(),
-        options.padding(),
-        options.output_padding(),
-        options.groups(),
-        options.dilation());
-  } else {
-    return torch::conv3d(
-        input,
-        weight,
-        bias,
-        options.stride(),
-        options.padding(),
-        options.dilation(),
-        options.groups());
+  if (c10::get_if<enumtype::kCircular>(&options.padding_mode())) {
+    std::vector<int64_t> expanded_padding = {
+      ((*options.padding())[2] + 1) / 2, (*options.padding())[2] / 2,
+      ((*options.padding())[1] + 1) / 2, (*options.padding())[1] / 2,
+      ((*options.padding())[0] + 1) / 2, (*options.padding())[0] / 2};
+    return F::conv3d(
+      F::pad(input, PadOptions(expanded_padding).mode(torch::kCircular)),
+      weight, bias,
+      Conv3dOptions()
```

Comment:
Why can't we just pass existing options instead recreating it?

Reply:
Thanks for the catch! Fixed.

Reply:
Sorry just realized that a call such as `options.padding(0)` is an in-place operation (i.e. it updates `padding_` of `options` instead of returning a copy of `options` that has `padding_` changed). For this reason, we will have to recreate it here.